// @fluentui/react-motions-preview/lib-commonjs/components/PresenceGroup.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "PresenceGroup", {
    enumerable: true,
    get: function() {
        return PresenceGroup;
    }
});
const _interop_require_wildcard = require("@swc/helpers/_/_interop_require_wildcard");
const _class_private_field_get = require("@swc/helpers/_/_class_private_field_get");
const _class_private_field_init = require("@swc/helpers/_/_class_private_field_init");
const _class_private_field_set = require("@swc/helpers/_/_class_private_field_set");
const _react = /*#__PURE__*/ _interop_require_wildcard._(require("react"));
const _getNextChildMapping = require("../utils/groups/getNextChildMapping");
const _getChildMapping = require("../utils/groups/getChildMapping");
const _PresenceGroupItemProvider = require("./PresenceGroupItemProvider");
var _mounted = /*#__PURE__*/ new WeakMap(), _handleExit = /*#__PURE__*/ new WeakMap();
class PresenceGroup extends _react.Component {
    static getDerivedStateFromProps(nextProps, { childMapping: prevChildMapping, firstRender }) {
        const nextChildMapping = (0, _getChildMapping.getChildMapping)(nextProps.children);
        return {
            childMapping: firstRender ? nextChildMapping : (0, _getNextChildMapping.getNextChildMapping)(prevChildMapping, nextChildMapping),
            firstRender: false
        };
    }
    componentDidMount() {
        (0, _class_private_field_set._)(this, _mounted, true);
    }
    componentWillUnmount() {
        (0, _class_private_field_set._)(this, _mounted, false);
    }
    render() {
        return /*#__PURE__*/ _react.createElement(_react.Fragment, null, Object.entries(this.state.childMapping).map(([childKey, childProps])=>/*#__PURE__*/ _react.createElement(_PresenceGroupItemProvider.PresenceGroupItemProvider, {
                ...childProps,
                childKey: childKey,
                key: childKey,
                onExit: (0, _class_private_field_get._)(this, _handleExit)
            }, childProps.element)));
    }
    constructor(props, context){
        super(props, context);
        (0, _class_private_field_init._)(this, _mounted, {
            writable: true,
            value: false
        });
        (0, _class_private_field_init._)(this, _handleExit, {
            writable: true,
            value: (childKey)=>{
                const currentChildMapping = (0, _getChildMapping.getChildMapping)(this.props.children);
                if (childKey in currentChildMapping) {
                    return;
                }
                if ((0, _class_private_field_get._)(this, _mounted)) {
                    this.setState((state)=>{
                        const childMapping = {
                            ...state.childMapping
                        };
                        delete childMapping[childKey];
                        return {
                            childMapping
                        };
                    });
                }
            }
        });
        this.state = {
            childMapping: {},
            firstRender: true
        };
    }
}
